src/js/tools/sentryMetroSerializer.ts

```diff
@@ -0,0 +1,256 @@
+};
+
+const createDebugIdModule = (debugId: string): Module<VirtualJSOutput> => {
+  const debugIdCode = createDebugIdSnippet(debugId);
+
+  return {
+    dependencies: new Map(),
+    getSource: () => Buffer.from(debugIdCode),
+    inverseDependencies: new CountingSet(),
+    path: DEBUG_ID_MODULE_PATH,
+    output: [
+      {
+        type: 'js/script/virtual',
+        data: {
+          code: debugIdCode,
+          lineCount: countLines(debugIdCode),
+          map: [],
+        },
+      },
+    ],
+  };
+};
+
+/**
+ * Deterministically hashes a string and turns the hash into a uuid.
+ * https://github.com/getsentry/sentry-javascript-bundler-plugins/blob/58271f1af2ade6b3e64d393d70376ae53bc5bd2f/packages/bundler-plugin-core/src/utils.ts#L174
+ */
+function stringToUUID(str: string): string {
+  const md5sum = crypto.createHash('md5');
+  md5sum.update(str);
+  const md5Hash = md5sum.digest('hex');
+
+  // Position 16 is fixed to either 8, 9, a, or b in the uuid v4 spec (10xx in binary)
+  // RFC 4122 section 4.4
+  const v4variant = ['8', '9', 'a', 'b'][md5Hash.substring(16, 17).charCodeAt(0) % 4] as string;
+
+  return `${md5Hash.substring(0, 8)}-${md5Hash.substring(8, 12)}-4${md5Hash.substring(
+    13,
+    16,
+  )}-${v4variant}${md5Hash.substring(17, 20)}-${md5Hash.substring(20)}`.toLowerCase();
+}
+
+const calculateDebugId = (bundle: Bundle): string => {
+  const hash = crypto.createHash('md5');
+  hash.update(bundle.pre);
+  for (const [, code] of bundle.modules) {
+    hash.update(code);
+  }
+  hash.update(bundle.post);
+
+  const debugId = stringToUUID(hash.digest('hex'));
+  return debugId;
+};
+
+const injectDebugId = (code: string, debugId: string): string =>
+  code.replace(new RegExp(DEBUG_ID_PLACE_HOLDER, 'g'), debugId);
+
+/**
+ * This function is expected to be called after serializer creates the final bundle object
+ * and before the source maps are generated.
+ *
+ * It injects a debug ID into the bundle and returns the modified bundle.
+ *
+ * Access it via `options.sentryBundleCallback` in your custom serializer.
+ */
+const sentryBundleCallback = (bundle: Bundle): Bundle => {
+  const debugId = calculateDebugId(bundle);
+  bundle.pre = injectDebugId(bundle.pre, debugId);
+  return bundle;
+};
+
+/**
+ * Creates the default Metro plain bundle serializer.
+ * This is used when the user does not provide a custom serializer.
+ */
+const createDefaultMetroSerializer = (serializerConfig: Partial<SerializerConfigT>): MetroSerializer => {
+  return (entryPoint, preModules, graph, options) => {
+    const createModuleId = options.createModuleId;
+    for (const module of graph.dependencies.values()) {
+      options.createModuleId(module.path);
+    }
+    const modules = [...graph.dependencies.values()].sort((a, b) => createModuleId(a.path) - createModuleId(b.path));
```

Comment:
l: Is there a way we can avoid the double iteration through the dependencies (once in for loop, once to construct final array?)

Reply:
Considering we combine `preModules` and `modules` below, do we need to sort all of them together?

Reply:
We have to sort the modules, otherwise, the source map would be broken, but I checked `baseJSBundle` and we can avoid the module id creation as it's be done in `baseJSBundle`.